File: Leetcode/Tree/BalancedBinaryTree.java
public class BalancedBinaryTree {
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode() {
        }
        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }
    public boolean isBalanced(TreeNode root) {
        return height(root) != -1;
    }
    public int height(TreeNode root){
        if(root == null) {
            return 0;
        }
        int lh = height(root.left);
        int rh = height(root.right);
        if((lh == -1) || (rh == -1)) return -1;
        if(Math.abs(lh - rh) > 1) return -1;
        return 1 + Math.max(lh,rh);
    }
}

// Current = 1
// Left = F(1) => -1
// Right = F(1)
// left = -1 => return -1;
// Current = 2
// Left = F(2) => 2
// Right = F(2) => 0
// => return -1;
// 	Current = 3
// Left = F(3) => 1
// Right = F(3) => 0
// => return 2
// Current = 4
// Left = F(4) => 0
// Right = F(4) => 0
// => return 1
// Current = NULL -> return 0;

File: Leetcode/300 Problems/Topics/Project1/BinaryTree/BalancedBinaryTree.cpp
// https://leetcode.com/problems/balanced-binary-tree/description/
#include <bits/stdc++.h>
using namespace std;
struct TreeNode
{
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

// Balanced BT -> for every node, |height(left) - height(right)| <= 1
int countDepth(TreeNode *root)
{
  if (root == NULL)
    return 0;
  int left = countDepth(root->left);

  // if left subtree is not balanced tree -> then return -1 => don't need to visit right subtree
  if (left == -1)
    return -1;

  int right = countDepth(root->right);

  // if right subtree is not balanced tree -> then return -1 => don't need to visit left subtree
  if (right == -1)
    return -1;

  if (abs(left - right) > 1)
    return -1;
  return 1 + max(left, right);
}
bool isBalanced(TreeNode *root)
{
  if (root == NULL)
    return true;

  return countDepth(root) != -1;
}